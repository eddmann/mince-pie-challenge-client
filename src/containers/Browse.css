.Browse {
  margin-top: 2em;
  margin-bottom: 2em;
}

.Browse .Pie {
  max-width: 266px !important;
}

.Browse .Pie .Rating {
  margin-left: -3px;
}
