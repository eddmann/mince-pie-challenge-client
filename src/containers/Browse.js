import React, { Component } from 'react';
import { connect } from 'react-redux'
import { push } from 'react-router-redux';
import { Loader, Card, Rating, Image, Message } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import { fetchPies } from '../actions/pie';

import DummyImage from './DummyImage.png';
import './Browse.css';

class Browse extends Component {
  static propTypes = {
    isFetching: PropTypes.bool.isRequired,
    pies: PropTypes.array.isRequired,
    browseUrl: PropTypes.string.isRequired,
    error: PropTypes.string,
  };

  componentDidMount() {
    this.props.dispatch(fetchPies(this.props.browseUrl));
  }

  onViewPie = (e, { id }) => {
    this.props.dispatch(push(`/pies/${id}`));
  };

  render() {
    if (this.props.isFetching) {
      return (
        <div className="Browse">
          <Loader active inline='centered' />
        </div>
      );
    }

    return (
      <div className="Browse">
        {this.props.error && <Message error>{this.props.error}</Message>}
        <Card.Group itemsPerRow={4} doubling>
          {this.props.pies.map(pie =>
            <Card className="Pie" key={pie.id} fluid onClick={this.onViewPie} id={pie.id}>
              <Image src={pie.thumbnail || DummyImage} />
              <Card.Content>
                <Card.Header content={pie.name} />
                <Card.Description>
                  <Rating className="Rating" disabled defaultRating={pie.rating.avg} maxRating={5} />
                </Card.Description>
              </Card.Content>
            </Card>
          )}
        </Card.Group>
      </div>
    );
  }
}

const mapStateToProps = ({ urls, browse, pies }) => ({
  isFetching: browse.isFetching,
  pies: browse.listing.map(id => pies[id]),
  browseUrl: urls.browseUrl,
  error: browse.error,
});

export default connect(mapStateToProps)(Browse);
