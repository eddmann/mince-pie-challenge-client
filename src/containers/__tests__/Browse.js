import React from 'react';
import { Browse } from '../Browse';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

const ID = '1234';
const NAME = 'Mince Pie';
const BROWSE_URL = 'browse.url';

it('fetches mince pies', () => {
  const fetchPies = jest.fn();
  const wrapper = shallow(<Browse
    fetchPies={fetchPies}
    push={jest.fn()}
    isFetching={true}
    pies={[]}
    browseUrl={BROWSE_URL}
    error={undefined}
  />);

  expect(fetchPies).toBeCalledWith(BROWSE_URL);
  expect(wrapper.find('Loader').length).toEqual(1);
});

it('lists mince pies', () => {
  const wrapper = shallow(<Browse
    fetchPies={jest.fn()}
    push={jest.fn()}
    isFetching={false}
    pies={[
      {
        id: ID,
        name: NAME,
        rating: { avg: 0, total: 0 },
        thumbnail: 'thumbnail.url'
      }
    ]}
    browseUrl={BROWSE_URL}
    error={undefined}
  />);

  expect(wrapper.find('Card').length).toEqual(1);
  expect(wrapper.find('Card')).toMatchSnapshot();
});

it('links to mince pie', () => {
  const push = jest.fn();
  const wrapper = shallow(<Browse
    fetchPies={jest.fn()}
    push={push}
    isFetching={false}
    pies={[
      {
        id: ID,
        name: NAME,
        rating: { avg: 0, total: 0 },
        thumbnail: 'thumbnail.url'
      }
    ]}
    browseUrl={BROWSE_URL}
    error={undefined}
  />);

  wrapper.find('Card').simulate('click', {}, { id: ID });

  expect(push).toBeCalledWith(`/pies/${ID}`);
});
