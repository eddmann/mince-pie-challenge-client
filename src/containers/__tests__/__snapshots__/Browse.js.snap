// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lists mince pies 1`] = `
<Card
  className="Pie"
  fluid={true}
  id="1234"
  key="1234"
  onClick={[Function]}
>
  <Image
    as="img"
    src="thumbnail.url"
    ui={true}
  />
  <CardContent>
    <CardHeader
      content="Mince Pie"
    />
    <CardDescription>
      <Rating
        className="Rating"
        clearable="auto"
        defaultRating={0}
        disabled={true}
        maxRating={5}
      />
    </CardDescription>
  </CardContent>
</Card>
`;
